lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -0,0 +1,390 @@
+        // both the `default` successor of a `switch` and one of its standard
+        // case
+        CaseIt = Switch->removeCase(CaseIt);
+      } else {
+        ++CaseIt;
+      }
+    }
+
+    // It should never be the case that we end up with a `switch` having only
+    // `PlaceHolderTarget` as its successor
+    if (Switch->getNumCases() == 0
+        and Switch->getDefaultDest() == PlaceHolderTarget) {
+      revng_abort();
+    }
+  }
+}
+
+/// We define a `Descriptor` for the information needed to identify a
+/// `Divergence`. Specifically, the divergence is composed by a set of divergent
+/// edges, and by the exit node which is divergent wrt. those divergent exits.
+using DivergentEdgesDescriptor = std::pair<BasicBlock *,
+                                           SmallSet<BasicBlock *, 4>>;
+using DivergenceDescriptorT = std::pair<DivergentEdgesDescriptor, BasicBlock *>;
+
+/// Helper function that tries to identify a `DivergenceDescriptor`
+static std::optional<DivergenceDescriptorT>
+electDivergence(BasicBlock *Candidate,
+                BasicBlock *Exit,
+                DomTreeOnView<BasicBlock, Scope> &DomTree) {
+
+  // We employ this `std::optional` as return value, in order to signal if we
+  // identify a candidate for IDS here
+  std::optional<DivergenceDescriptorT> DivergenceDescriptor;
+
+  // Check if it is the conditional making it a divergent node
+  if (isConditional(Candidate) and DomTree.dominates(Candidate, Exit)) {
+    using ScopeBlock = Scope<BasicBlock *>;
+
+    SetVector<BasicBlock *> Successors;
+    for (BasicBlock *Successor : children<ScopeBlock>(Candidate)) {
+      Successors.insert(Successor);
+    }
+
+    std::map<BasicBlock *, SmallSet<BasicBlock *, 4>> ReachablesFromSuccessor;
+    size_t SuccessorsSize = std::distance(Successors.begin(), Successors.end());
+
+    for (auto *Successor : Successors) {
+      SmallSet<BasicBlock *, 4> ReachableExits;
+
+      // Explore all the exits that are reachable from each Successor.
+      // For having a divergent exit, we need to find a set of
+      // successors that reach only the exit under analysis
+      for (auto *DFSNode : depth_first(Scope<BasicBlock *>(Successor))) {
+        if (isExit(DFSNode)) {
+          ReachableExits.insert(DFSNode);
+        }
+      }
+
+      ReachablesFromSuccessor[Successor] = ReachableExits;
+    }
+
+    SmallSet<BasicBlock *, 4> DivergentSuccessors;
+    for (const auto &[Successor, ReachableExits] : ReachablesFromSuccessor) {
+      if (ReachableExits.size() == 1 and *ReachableExits.begin() == Exit) {
+        DivergentSuccessors.insert(Successor);
+      } else if (ReachableExits.contains(Exit)) {
+
+        // If we reach `Exit` from a successor which is not a candidate for
+        // being divergent, it means that `Exit` is reached also by a non
+        // divergent successor, and this contradicts the definition, so we
+        // cannot find a divergence here
+        return DivergenceDescriptor;
```

Comment:
Use `return std::nullopt` here, otherwise is not obvious that here it's always `nullopt`. There's the comment, but comments go out of sync.

Reply:
Right, changed this here https://github.com/revng/revng/blob/a3b1cbac550057e016f37c8f9eb201e8ab290b82/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L214